execution-engine/engine-core/src/execution/runtime/mod.rs

```diff
@@ -560,6 +575,577 @@ where
+                    match res {
+                        Ok(key) => Ok(key.into_uref().into_iter().collect()),
+                        Err(_) => Ok(vec![]),
+                    }
+                }
+                (CLType::Key, CLType::U64) => {
+                    let res: Result<Key, u64> = value.to_owned().into_t()?;
+                    match res {
+                        Ok(key) => Ok(key.into_uref().into_iter().collect()),
+                        Err(_) => Ok(vec![]),
+                    }
+                }
+                (CLType::Key, CLType::U128) => {
+                    let res: Result<Key, U128> = value.to_owned().into_t()?;
+                    match res {
+                        Ok(key) => Ok(key.into_uref().into_iter().collect()),
+                        Err(_) => Ok(vec![]),
+                    }
+                }
+                (CLType::Key, CLType::U256) => {
+                    let res: Result<Key, U256> = value.to_owned().into_t()?;
+                    match res {
+                        Ok(key) => Ok(key.into_uref().into_iter().collect()),
+                        Err(_) => Ok(vec![]),
+                    }
+                }
+                (CLType::Key, CLType::U512) => {
+                    let res: Result<Key, U512> = value.to_owned().into_t()?;
+                    match res {
+                        Ok(key) => Ok(key.into_uref().into_iter().collect()),
+                        Err(_) => Ok(vec![]),
+                    }
+                }
+                (CLType::Key, CLType::Unit) => {
+                    let res: Result<Key, ()> = value.to_owned().into_t()?;
+                    match res {
+                        Ok(key) => Ok(key.into_uref().into_iter().collect()),
+                        Err(_) => Ok(vec![]),
+                    }
+                }
+                (CLType::Key, CLType::String) => {
+                    let res: Result<Key, String> = value.to_owned().into_t()?;
+                    match res {
+                        Ok(key) => Ok(key.into_uref().into_iter().collect()),
+                        Err(_) => Ok(vec![]),
+                    }
+                }
+                (CLType::Key, CLType::URef) => {
+                    let res: Result<Key, URef> = value.to_owned().into_t()?;
+                    match res {
+                        Ok(key) => Ok(key.into_uref().into_iter().collect()),
+                        Err(uref) => Ok(vec![uref]),
+                    }
+                }
+                (CLType::Key, CLType::Key) => {
+                    let res: Result<Key, Key> = value.to_owned().into_t()?;
+                    match res {
+                        Ok(key) => Ok(key.into_uref().into_iter().collect()),
+                        Err(key) => Ok(key.into_uref().into_iter().collect()),
+                    }
+                }
+                (_, _) => Ok(vec![]),
+            },
+            CLType::Map { key, value: val } => match (&**key, &**val) {
```

Comment:
Shouldn't we also be matching on all `(xxx, CLType::URef/Key)`?

The same also probably applies to `Tuple2` for the second element, but maybe isn't applicable for `Result`?  For `Result<xxx, URef/Key>` it's maybe reasonable to ignore `URef/Key`s passed in the `Err` variant?  If so, then I guess we shouldn't handle _any_ such cases (e.g. for `Result<Key, URef>` we're returning the `URef`)

Reply:
Oh, I totally missed this :grimacing: I'd say we should probably do this for tuples and Results, too - or this will be an opening for passing in unchecked URefs (not that there aren't any such openings at the moment, anyway...)